cfg/config.go

```diff
@@ -4,27 +4,111 @@ import (
 	"encoding/json"
 	"io/ioutil"
 	"os"
+	"path/filepath"
 
 	"github.com/muesli/beehive/bees"
+	gap "github.com/muesli/go-app-paths"
+	log "github.com/sirupsen/logrus"
 )
 
+const appName = "beehive"
+
+var cfgFileName = "beehive.conf"
+var lookupFunc = Lookup
+
 // Config contains an entire configuration set for Beehive
 type Config struct {
 	Bees    []bees.BeeConfig
 	Actions []bees.Action
 	Chains  []bees.Chain
 }
 
+// DefaultPath returns Beehive's default config path
+//
+// The path returned is OS dependant. If there's an error
+// while trying to figure out the OS dependant path, "beehive.conf"
+// in the current working dir is returned.
+func DefaultPath() string {
+	userScope := gap.NewScope(gap.User, appName)
+	path, err := userScope.ConfigPath(cfgFileName)
+	if err != nil {
+		return cfgFileName
+	}
+
+	return path
+}
+
+// Lookup an existing user config file
+//
+// Returns an empty array when no configs are found.
+func Lookup() []string {
+	paths := []string{}
+	defaultPath := DefaultPath()
+	if Exist(defaultPath) {
+		paths = append(paths, defaultPath)
+	}
+
+	// Prepend ./beehive.conf to the search path if exists, takes priority
+	// over the rest
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Errorf("Error getting current working directory. err: %v", err)
+		cwd = "."
+	}
+	cwdCfg := filepath.Join(cwd, cfgFileName)
+	if Exist(cwdCfg) {
+		paths = append([]string{cwdCfg}, paths...)
+	}
+
+	return paths
+}
+
+// FindUserConfigPath tries to find the config file.
+//
+// If a config file is found in the current working directory, that's returned.
+// Otherwise we try to locate it following an OS dependant:
+//
+// Unix:
+//   - ~/.config/app/filename.conf
+// macOS:
+//   - ~/Library/Preferences/app/filename.conf
+// Windows:
+//   - %LOCALAPPDATA%/app/Config/filename.conf
+//
+// If no valid config file is found, an empty string is returned.
+func FindUserConfigPath() string {
+	paths := lookupFunc()
+	if len(paths) == 0 {
+		return ""
+	}
+	return paths[0]
+}
+
+// Load tries to load the default config file, returning an empty config object
+// if not found
+//
+// Returns the path of the config found and the Config object, returning an error
+// also if loading the config from the filesystem fails.
+func Load() (string, Config, error) {
```

Comment:
I think the naming is a bit confusing here. Maybe `Load` should be called `LoadDefaultConfig` to differentiate it from `LoadConfig` a bit more. Part of me thinks there should only be one public method responsible for loading the config. Maybe the caller should retrieve the config path first and then only call `LoadConfig`?

This is probably awfully nitpicky.

Reply:
Make sense to me too. I'm getting rid of `LoadConfig` and replacing it with a single `cfg.Load` public function. The `Config` suffix was also a bit repetitive given the package is also named `cfg`.